runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -12056,3 +12071,319 @@ J9::Z::TreeEvaluator::tabortEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+      }
+
+   TR::LabelSymbol* callLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol* callEndLabel = generateLabelSymbol(cg);
+   TR::Instruction *gcPoint = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, callLabel);
+   gcPoint->setNeedsGCMap(0);
+   auto readBarHelperSnippet = new (cg->trHeapMemory()) TR::S390HelperCallSnippet(cg, node, callLabel,
+                                                                                  cg->symRefTab()->findOrCreateRuntimeHelper(TR_softwareReadBarrier, false, false, false),
+                                                                                  callEndLabel);
+   cg->addSnippet(readBarHelperSnippet);
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, callEndLabel);
+
+   // Reload the object after helper call.
+   generateRXInstruction(cg, loadOpCode, node, resultReg, generateS390MemoryReference(fieldAddrReg, 0, cg));
+   TR::Instruction* cursor = generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, doneLabel);
+   if (notInsideICF)
+      {
+      cursor->setDependencyConditions(deps);
+      doneLabel->setEndInternalControlFlow();
+      }
+
+   // produce decompressed value in the end
+   if (shouldShift)
+      {
+      generateRSInstruction(cg, TR::InstOpCode::SLLG, node, resultReg, resultReg, shiftAmount);
+      }
+
+   cg->generateDebugCounter(TR::DebugCounter::debugCounterName(cg->comp(), "readBar/total"), 1, TR::DebugCounter::Cheap);
+   cg->stopUsingRegister(fieldAddrReg);
+
+   return resultReg;
+   }
+
+TR::Register *
+J9::Z::TreeEvaluator::arraycopyEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+   {
+   if (node->isReferenceArrayCopy())
+      {
+      TR::TreeEvaluator::referenceArraycopyEvaluator(node, cg);
+      }
+   else
+      {
+      OMR::TreeEvaluatorConnector::arraycopyEvaluator(node, cg);
+      }
+   return NULL;
+   }
+
+TR::Register *
+J9::Z::TreeEvaluator::referenceArraycopyEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+   {
+   TR::Node* byteSrcObjNode = node->getChild(0);
+   TR::Node* byteDstObjNode = node->getChild(1);
+   TR::Node* byteSrcNode    = node->getChild(2);
+   TR::Node* byteDstNode    = node->getChild(3);
+   TR::Node* byteLenNode    = node->getChild(4);
+
+   TR::Register* byteSrcObjReg = cg->evaluate(byteSrcObjNode);
+   TR::Register* byteDstObjReg = cg->evaluate(byteDstObjNode);
+   TR::Register* byteSrcReg = NULL;
+   TR::Register* byteDstReg = NULL;
+   TR::Register* byteLenReg = NULL;
+
+   if (!node->chkNoArrayStoreCheckArrayCopy())
+      {
+      byteSrcReg = cg->evaluate(byteSrcNode);
+      byteDstReg = cg->evaluate(byteDstNode);
+      byteLenReg = cg->evaluate(byteLenNode);
+
+      genArrayCopyWithArrayStoreCHK(node, byteSrcObjReg, byteDstObjReg, byteSrcReg, byteDstReg, byteLenReg, cg);
+
+      cg->decReferenceCount(byteSrcNode);
+      cg->decReferenceCount(byteDstNode);
+      cg->decReferenceCount(byteLenNode);
```

Comment:
Right, on the surface,  array copy evaluator helpers are not decrementing node reference counts consistently. However, the general rule is that whoever calls `cg->evaluate(node)` will need to decrement that node's ref count.

The three nodes you are referring to are not decremented in the `else` block because `primitiveArraycopyEvaluator` in the `else` block handles `evaluate-and-decrement` already.

Reply:
There are 5 nodes and the other two also follow the `evaluate-decrement` rule.

Reply:
Ah yes, thanks for explaining that. I was under the impression that the last two children are the object references, but it's actually the first two! You're right everything is working as intended.